### Preamble

In [1]:
import matplotlib.pyplot as plt
import numpy as np
from trapInt import funct, trap, trapIntMain 

0.00012071080015974012
1.0973699009486637e-05
5.748127799298786e-06
3.893892953410472e-06
2.9441629314508333e-06
2.3668760840145353e-06
1.978863622742466e-06
1.7001503740399215e-06
1.4902552907258837e-06
1.3264910354582327e-06
1.1951552907341068e-06
1.0874835905116953e-06
9.976088803931304e-07
9.214555585224105e-07
8.561041711097062e-07
7.994084818552238e-07
7.497557561464419e-07
7.059104956130824e-07
6.669098127076722e-07
6.319930814235875e-07
6.005505852613169e-07
5.720885576376844e-07
5.462022005797706e-07
5.225571035882837e-07
5.008742760548301e-07
4.809189433352401e-07
4.6249303031764555e-07
4.4542686628012973e-07
4.29575347959954e-07
4.1481351287231405e-07
4.010321112968285e-07
3.88137218365614e-07
3.7604563707382965e-07
3.6468481701135147e-07
3.539902808658013e-07
3.4390511705062465e-07
3.3437858859564074e-07
3.253657292479062e-07
3.1682594081594573e-07
3.087229369924671e-07
3.0102411252563854e-07
2.936998917402178e-07
2.867236768011594e-07
2.800711322858973e-07
2.73720528687349

KeyboardInterrupt: 

### Simpson's Rule

In [ ]:
def simp(n, min, max):
    if n % 2 == 0:
        print("Error: n must be odd")
        return None
        
    dx = (max - min) / (n-1)
    sum = 0.0

    for i in range(1, n-1):
        x = min + i*dx
        if i % 2 == 0:
            sum += 2*funct(x)
        else:
            sum += 4*funct(x)
    
    # Add the endpoints
    sum += funct(min) + funct(max)
    sum *= dx/3.0
    

    nextN = 2*n + 1
    newDeltaX = (max - min) / (2*n)
    sumNextN = 0

    for i in range(nextN):
        x = min + i*newDeltaX

    return sum

### Testing Simpson's Rule function

In [ ]:
a = 1.0
b = 2.0
exact = (2/3) * (b ** (3/2) - a ** (3/2))

'''
for i in range(3, 10000, 2):
    approx = simp(i, a, b)
    error = abs(exact - approx)
    print('{:6d}{:.12e}'.format(i, error))
for i in range(1001, 3000000, 10000):
    approx = simp(i, a, b)
    error = abs(exact - approx)
    print('{:6d}{:.12e}'.format(i, error))
''' 


## Error plots


# Regression Polynomials 

Only the approximation dominate region of the trapezoid rule data can be fit due to the large values of $N$ required to calculate the round-off dominant region. The trapezoid rule's approximation region produced the following polynomial: $y=1.2383 \cdot x^{-1.003}$. Simpson's rule data could be fit both in the approximation and round-off dominate regions producing the following polynomials respectively: $y=0.0025991\cdot x^{-4.0719}$ and $1.5489 \times 10^{-17} \cdot x^{0.51375}$. 

# Optimum N for least total error
Optimum values of N may be determined by manual inspection of the xmgrace plots.  The optimum N occurs at minima of the trapezoid and Simpson rules error data. Using this method, $N\approx 10^6$ for the trapezoid rule and $N\approx 10^{3.5}$ for Simpson's rule. These values can be determined analytically using Equations 5.23 and 5.25.

# Graduate Problem

In [ ]:
trapErrorEst = abs( / )
simpErrorEst = abs( / )